src/i2c.rs

```diff
@@ -1,579 +1,1819 @@
+        type Error = I2cError;
     }
 
-    fn transaction(
-        &mut self,
-        address: u8,
-        operations: &mut [embedded_hal::i2c::Operation<'_>],
-    ) -> Result<(), Self::Error> {
-        I2cDriver::transaction(self, address, operations, BLOCK).map_err(to_i2c_err)
+    #[cfg(not(esp_idf_i2c_isr_iram_safe))]
+    impl<'d> embedded_hal_async::i2c::I2c<embedded_hal::i2c::SevenBitAddress>
+        for OwnedAsyncI2cDeviceDriver<'d>
+    {
+        async fn read(&mut self, _address: u8, buffer: &mut [u8]) -> Result<(), Self::Error> {
+            Self::read(self, buffer, BLOCK).await.map_err(to_i2c_err)
+        }
+
+        async fn write(&mut self, _address: u8, bytes: &[u8]) -> Result<(), Self::Error> {
+            Self::write(self, bytes, BLOCK).await.map_err(to_i2c_err)
+        }
+
+        async fn write_read(
+            &mut self,
+            _address: u8,
+            bytes: &[u8],
+            buffer: &mut [u8],
+        ) -> Result<(), Self::Error> {
+            Self::write_read(self, bytes, buffer, BLOCK)
+                .await
+                .map_err(to_i2c_err)
+        }
+
+        async fn transaction(
+            &mut self,
+            _address: u8,
+            _operations: &mut [Operation<'_>],
+        ) -> Result<(), Self::Error> {
+            unimplemented!("transactional not implemented")
+        }
     }
-}
 
-fn to_i2c_err(err: EspError) -> I2cError {
-    if err.code() == ESP_FAIL {
-        I2cError::new(ErrorKind::NoAcknowledge(NoAcknowledgeSource::Unknown), err)
-    } else {
-        I2cError::other(err)
+    #[cfg(not(esp32c2))]
+    pub struct I2cSlaveDriver<'d> {
+        i2c: u8,
+        handle: i2c_slave_dev_handle_t,
+        _p: PhantomData<&'d mut ()>,
     }
-}
 
-#[cfg(not(esp32c2))]
-pub struct I2cSlaveDriver<'d> {
-    i2c: u8,
-    _p: PhantomData<&'d mut ()>,
-}
+    #[cfg(not(esp32c2))]
+    unsafe impl<'d> Send for I2cSlaveDriver<'d> {}
+
+    #[cfg(not(esp32c2))]
+    impl<'d> I2cSlaveDriver<'d> {
+        pub fn new<I2C: I2c>(
+            _i2c: impl Peripheral<P = I2C> + 'd,
+            sda: impl Peripheral<P = impl InputPin + OutputPin> + 'd,
+            scl: impl Peripheral<P = impl InputPin + OutputPin> + 'd,
+            address: config::DeviceAddress,
+            config: &config::SlaveDeviceConfig,
+        ) -> Result<Self, EspError> {
+            let handle = init_slave_device(_i2c, sda, scl, address, config)?;
+
+            enable_slave_isr_callback(handle, I2C::port() as _)?;
+
+            Ok(Self {
+                i2c: I2C::port() as _,
+                handle,
+                _p: PhantomData,
+            })
+        }
+
+        pub fn read(&mut self, buffer: &mut [u8], _timeout: TickType_t) -> Result<usize, EspError> {
+            esp!(unsafe { i2c_slave_receive(self.handle, buffer.as_mut_ptr(), buffer.len()) })?;
 
-#[cfg(not(esp32c2))]
-unsafe impl<'d> Send for I2cSlaveDriver<'d> {}
+            todo!("How to block?");
```

Comment:
Roughly speaking (there are more details to this):
You need a blocking mutex and a condvar. You need to lock the blocking mutex and then wait on the condvar. Waiting on the condvar will temporarily unlock the mutex. In the callback handler, besides notifying the static notification, you also need to lock the mutex and then notify the condvar.